File: iosApp/iosApp/Modules/Confirmation/ConfirmationInteractor.swift
// Created by web3d4v on 10/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import web3lib

protocol ConfirmationInteractor {
    func send(
        addressTo: String,
        network: Network,
        currency: Currency,
        amount: BigInt,
        fee: Web3NetworkFee,
        password: String,
        salt: String,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    )
    func sendNFT(
        addressFrom: String,
        addressTo: String,
        network: Network,
        nft: NFTItem,
        password: String,
        salt: String,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    )
     func castVote(
        proposalId: String,
        support: Bool,
        password: String,
        salt: String,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    )
    func executeSwap(
        network: Network,
        password: String,
        salt: String,
        swapService: UniswapService,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    )
}

final class DefaultConfirmationInteractor {
    private let walletService: WalletService
    private let nftsService: NFTsService

    init(
        walletService: WalletService,
        nftsService: NFTsService
    ) {
        self.walletService = walletService
        self.nftsService = nftsService
    }
}

extension DefaultConfirmationInteractor: ConfirmationInteractor {
    
    func send(
        addressTo: String,
        network: Network,
        currency: Currency,
        amount: BigInt,
        fee: Web3NetworkFee,
        password: String,
        salt: String,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    ) {
        do {
            try walletService.unlock(
                password: password,
                salt: salt,
                network: network
            )
            walletService.transfer(
                to: addressTo,
                currency: currency,
                amount: amount,
                network: network,
                completionHandler: { response, error in
                    if let error = error {
                        handler(.failure(error))
                        return
                    }
                    guard let response = response else {
                        handler(.failure(ConfirmationInteractorError.noResponse))
                        return
                    }
                    handler(.success(response))
                }
            )
        } catch {
            handler(.failure(error))
        }
    }

    func sendNFT(
        addressFrom: String,
        addressTo: String,
        network: Network,
        nft: NFTItem,
        password: String,
        salt: String,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    ) {
        guard let tokenIdInt = try? nft.tokenId.int() else {
            handler(.failure(ConfirmationInteractorError.failedToParseTokenId))
            return
        }
        do {
            try walletService.unlock(
                password: password,
                salt: salt,
                network: network
            )
            let contract = ERC721(address: Address.HexString(hexString: nft.address))
            walletService.contractSend(
                contractAddress: contract.address.hexString,
                data: contract.transferFrom(
                    from: Address.HexString(hexString: addressFrom),
                    to: Address.HexString(hexString: addressTo),
                    tokenId: BigInt.Companion().from(long: Int64(tokenIdInt))
                ),
                network: network,
                completionHandler: { response, error in
                    if let error = error {
                        handler(.failure(error))
                        return
                    }
                    guard let response = response else {
                        handler(.failure(ConfirmationInteractorError.noResponse))
                        return
                    }
                    self.nftsService.nftSent(identifier: nft.identifier)
                    handler(.success(response))
                }
            )
        } catch {
            handler(.failure(error))
        }
    }
    
    func castVote(
        proposalId: String,
        support: Bool,
        password: String,
        salt: String,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    ) {
        do {
            guard let id = try? proposalId.int() else {
                handler(.failure(ConfirmationInteractorError.failedProposalId))
                return
            }
            let network = Network.ethereum()
            try walletService.unlock(
                password: password,
                salt: salt,
                network: network
            )
            let contract = CultGovernor()
            let supportInt = UInt32(support ? 1 : 0)
            walletService.contractSend(
                contractAddress: contract.address.hexString,
                data: contract.castVote(proposalId: UInt32(id), support: supportInt),
                network: network,
                completionHandler:  { response, error in
                    if let error = error {
                        handler(.failure(error))
                        return
                    }
                    guard let response = response else {
                        handler(.failure(ConfirmationInteractorError.noResponse))
                        return
                    }
                    handler(.success(response))
                }
            )
        } catch {
            handler(.failure(error))
        }
    }
    
    func executeSwap(
        network: Network,
        password: String,
        salt: String,
        swapService: UniswapService,
        handler: @escaping (Result<TransactionResponse, Error>) -> Void
    ) {
        do {
            try walletService.unlock(
                password: password,
                salt: salt,
                network: network
            )
            swapService.executeSwap { response, error in
                if let error = error {
                    handler(.failure(error))
                    return
                }
                guard let response = response else {
                    handler(.failure(ConfirmationInteractorError.noResponse))
                    return
                }
                handler(.success(response))
            }
        } catch {
            handler(.failure(error))
        }
    }
}

enum ConfirmationInteractorError: Error {
    case noResponse
    case failedToParseTokenId
    case failedProposalId
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/confirmation/ConfirmationInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.confirmation

import com.sonsofcrypto.web3lib.provider.model.TransactionResponse
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.uniswap.UniswapService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.signer.contracts.CultGovernor
import com.sonsofcrypto.web3lib.signer.contracts.ERC721
import com.sonsofcrypto.web3lib.types.*
import com.sonsofcrypto.web3lib.types.Address.HexString
import com.sonsofcrypto.web3lib.utils.BigInt
import com.sonsofcrypto.web3walletcore.services.nfts.NFTItem
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService

interface ConfirmationInteractor {
    @Throws(Throwable::class)
    suspend fun send(
        input: ConfirmationWireframeContext.SendContext, password: String, salt: String,
    ): TransactionResponse

    @Throws(Throwable::class)
    suspend fun sendNFT(
        input: ConfirmationWireframeContext.SendNFTContext, password: String, salt: String,
    ): TransactionResponse

    @Throws(Throwable::class)
    suspend fun castVote(
        input: ConfirmationWireframeContext.CultCastVoteContext, password: String, salt: String,
    ): TransactionResponse

    @Throws(Throwable::class)
    suspend fun swap(
        network: Network, password: String, salt: String, swapService: UniswapService,
    ): TransactionResponse

    fun fiatPrice(currency: Currency): Double

    fun trackNFTAsSent(nftItem: NFTItem)
}

class DefaultConfirmationInteractor(
    private val walletService: WalletService,
    private val nftsService: NFTsService,
    private val currencyStoreService: CurrencyStoreService
): ConfirmationInteractor {

    @Throws(Throwable::class)
    override suspend fun send(
        input: ConfirmationWireframeContext.SendContext, password: String, salt: String,
    ): TransactionResponse {
        walletService.unlock(password, salt, input.network)
        return walletService.transfer(input.addressTo, input.currency, input.amount, input.network)
    }

    @Throws(Throwable::class)
    override suspend fun sendNFT(
        input: ConfirmationWireframeContext.SendNFTContext, password: String, salt: String,
    ): TransactionResponse {
        walletService.unlock(password, salt, input.network)
        val contract = ERC721(HexString(input.nftItem.address))
        val tx = walletService.contractSend(
            contract.address.hexString,
            contract.transferFrom(
                HexString(input.addressFrom), HexString(input.addressTo), input.nftItem.tokenId
            ),
            input.network
        )
        return tx
    }

    @Throws(Throwable::class)
    override suspend fun castVote(
        input: ConfirmationWireframeContext.CultCastVoteContext, password: String, salt: String,
    ): TransactionResponse {
        // TODO: Review this when adding support to send a NFT on a different network than Ethereum.
        val network = Network.ethereum()
        walletService.unlock(password, salt, network)
        val contract = CultGovernor()
        return walletService.contractSend(
            contract.address.hexString,
            contract.castVote(
                input.cultProposal.id.toUInt(), (if (input.approve) 1 else 0).toUInt()
            ),
            network
        )
    }

    @Throws(Throwable::class)
    override suspend fun swap(
        network: Network, password: String, salt: String, swapService: UniswapService,
    ): TransactionResponse {
        walletService.unlock(password, salt, network)
        return swapService.executeSwap()
    }

    override fun fiatPrice(currency: Currency): Double =
        currencyStoreService.marketData(currency)?.currentPrice ?: 0.toDouble()

    override fun trackNFTAsSent(nftItem: NFTItem) = nftsService.nftSent(nftItem.identifier)
}
